File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *next;
 *     struct Node *random;
 * };
 */

void insertbetween(struct Node*head){
    struct Node*temp=head;
    while(temp!=NULL){
        struct Node*n=(struct Node*)malloc(sizeof(struct Node));
        n->val=temp->val;
        n->next=temp->next;
        temp->next=n;
        temp=temp->next->next;
    }
}

void connectrandom(struct Node*head){
    struct Node*temp=head;
    while(temp!=NULL){
        struct Node*copynode=temp->next;
        if(temp->random!=NULL){
        copynode->random=temp->random->next;
        }
        else{
            copynode->random=NULL;
        }
        temp=temp->next->next;
    }
}

struct Node* connectnext(struct Node*head){
    struct Node*temp=head;
    struct Node*dummy=(struct Node*)malloc(sizeof(struct Node));
    dummy->next = NULL;
    struct Node*res=dummy;
    while(temp!=NULL){
        res->next=temp->next;
        temp->next=temp->next->next;
        res=res->next;
        temp=temp->next;
    }
    return dummy->next;
}

struct Node* copyRandomList(struct Node* head) {
    insertbetween(head);
    connectrandom(head);
    return connectnext(head);	
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if(head==null){
            return null;
        }
        Node cur=head;
        while(cur!=null){
            Node newone=new Node(cur.val);
            newone.next=cur.next;
            cur.next=newone;
            cur=newone.next;
        }
        
        cur=head;
        while(cur!=null){
            if(cur.random!=null){
                cur.next.random=cur.random.next;
            }
            cur=cur.next.next;
        }
        
        cur=head;
        Node newhead=head.next;
        Node newcur=newhead;
        while(cur != null){
            cur.next=newcur.next;
            cur=cur.next;
            if(cur!=null){
                newcur.next=cur.next;
                newcur=newcur.next;
            }  
        }
        return newhead;
    }
}